lib/veri.ml

```diff
@@ -0,0 +1,207 @@
+open Core_kernel.Std
+open Bap.Std
+open Bap_traces.Std
+module Bap_value = Value
+open Bil.Result
+open Monad.State
+open Trace
+
+let do_nothing () = return ()
+
+type point = {
+  code : Chunk.t;
+  side : event list
+}
+let is_first_read ctxt ev =
+  match Bap_value.get Event.register_read ev with
+  | None -> false
+  | Some mv -> Option.is_none (ctxt#lookup (Move.cell mv))
+
+let is_first_load ctxt ev = 
+  match Bap_value.get Event.memory_load ev with
+  | None -> false
+  | Some mv ->
+    match Seq.find ~f:(fun (v,_) -> v = ctxt#mem) ctxt#bindings with
+    | None -> true
+    | Some (_,s) ->
+      match value s with
+      | Bil.Mem storage -> Option.is_none (storage#load (Move.cell mv))
+      | _ -> true 
+
+let is_init_event ctxt ev =
+  is_first_read ctxt ev || is_first_load ctxt ev
+
+class context report arch = object(self:'s)
+  inherit Veri_traci.context arch
+  val report : Veri_report.t = report
+  val events : event list = []
+  val selves = None
+  val current = `Fst
+  val descr : string option = None
+
+  method split = 
+    let self = {< events = []; descr = None >} in
+    {< selves = Some (self, self); current = `Fst >}
+       
+  method register_event ev = match selves with
+    | None -> {< events = ev :: events >}
+    | Some (main, shadow) ->
+      match current with
+      | `Fst ->
+        let main = main#register_event ev in
+        {<selves = Some (main, shadow) >}
+      | `Snd ->
+        let shadow' = shadow#register_event ev in
+        {<selves = Some (main,shadow') >}
+
+  method events = events
+  method main = fst (Option.value_exn selves)
+  method shadow = snd (Option.value_exn selves)
+  method replay = {<current = `Snd; >}
+
+  method merge: 's = 
+    let is_exists_event ev evs = 
+      List.exists ~f:(fun ev' -> ev = ev') evs in   
+    let events = self#main#events in
+    let events' = self#shadow#events in
+    let init = List.filter ~f:(is_init_event self#main) events' in
```

Comment:
First of all, I don't understand why we need `init`. Second, you should move all the comparison logic away from the veri. Veri deligates it to the context. Basically, merge should be just a one-liner that calls ctxt merge, and do all the required reinition of the selves.